File: katas/typescript/src/CleanCode/format/format-good.ts
// Cumplimiento de Clean Code: Buen Formato
// ✅ Solución: Formato horizontal y vertical correcto

// ✅ Líneas cortas (idealmente max 80-120 caracteres)
// ✅ Espaciado apropiado entre operadores y elementos

interface Product {
  id: number;
  name: string;
  price: number;
  category: string;
  stock: number;
  supplier: string;
  warranty: string;
  discountedPrice?: number;
}

class ProductService {
  private products: Product[] = [
    {
      id: 1,
      name: "Laptop",
      price: 1200,
      category: "Electronics",
      stock: 50,
      supplier: "TechCorp",
      warranty: "2 years",
    },
    {
      id: 2,
      name: "Mouse",
      price: 25,
      category: "Accessories",
      stock: 200,
      supplier: "TechCorp",
      warranty: "1 year",
    },
  ];

  // ✅ Líneas cortas, una acción por línea
  public findProductByIdAndUpdateStockAndCalculateDiscountAndSendNotification(productId: number, quantity: number, discountRate: number): Product | null {
    const product = this.findProductById(productId);

    if (!product) {
      return null;
    }

    product.stock -= quantity;
    const discountedPrice = this.calculateDiscountedPrice(product.price, discountRate);

    console.log(`Producto ${product.name} actualizado. ` + `Stock: ${product.stock}, ` + `Precio con descuento: ${discountedPrice}`);

    return {
      ...product,
      discountedPrice,
    };
  }

  private findProductById(id: number): Product | undefined {
    return this.products.find((product) => product.id === id);
  }

  private calculateDiscountedPrice(price: number, discountRate: number): number {
    return price * discountRate;
  }
}

// ✅ Espaciado vertical apropiado
// ✅ Líneas en blanco separan conceptos relacionados
// ✅ Métodos relacionados están cerca

interface OrderItem {
  price: number;
  quantity: number;
}

interface Order {
  id: number;
  customerId: number;
  items: OrderItem[];
  total: number;
  date: Date;
  status: string;
}

class OrderProcessor {
  private static readonly DISCOUNT_THRESHOLD = 100;
  private static readonly DISCOUNT_RATE = 0.9;

  private orders: Order[] = [];

  // ✅ Grupo 1: Procesamiento de órdenes
  public processOrder(customerId: number, items: OrderItem[], paymentMethod: string): boolean {
    if (!this.isValidOrder(customerId, items)) {
      return false;
    }

    const total = this.calculateTotal(items);
    const order = this.createOrder(customerId, items, total);

    this.saveOrder(order);
    this.logOrderProcessed(total);

    return true;
  }

  private isValidOrder(customerId: number, items: OrderItem[]): boolean {
    return customerId > 0 && items && items.length > 0;
  }

  private calculateTotal(items: OrderItem[]): number {
    let total = items.reduce((sum, item) => sum + item.price * item.quantity, 0);

    if (total > OrderProcessor.DISCOUNT_THRESHOLD) {
      total = total * OrderProcessor.DISCOUNT_RATE;
    }

    return total;
  }

  private createOrder(customerId: number, items: OrderItem[], total: number): Order {
    return {
      id: Date.now(),
      customerId,
      items,
      total,
      date: new Date(),
      status: "pending",
    };
  }

  private saveOrder(order: Order): void {
    this.orders.push(order);
  }

  private logOrderProcessed(total: number): void {
    console.log(`Orden procesada: $${total}`);
  }

  // ✅ Grupo 2: Consultas de órdenes (separado por línea en blanco)
  public getOrders(): Order[] {
    return this.orders;
  }

  // ✅ Grupo 3: Cancelación de órdenes (separado por línea en blanco)
  public cancelOrder(orderId: number): boolean {
    const index = this.orders.findIndex((order) => order.id === orderId);

    if (index >= 0) {
      this.orders.splice(index, 1);
      return true;
    }

    return false;
  }
}

// ✅ Formato consistente en toda la clase
// ✅ Espaciado uniforme, indentación correcta

interface User {
  name: string;
  email: string;
  age: number;
  createdAt: Date;
}

class UserManager {
  private users: User[] = [];

  public addUser(name: string, email: string, age: number): User {
    const user: User = {
      name,
      email,
      age,
      createdAt: new Date(),
    };

    this.users.push(user);
    return user;
  }

  public findUser(email: string): User | undefined {
    return this.users.find((user) => user.email === email);
  }

  public deleteUser(email: string): void {
    this.users = this.users.filter((user) => user.email !== email);
  }
}

// Uso - También con buen formato
console.log("=== Cumplimiento de Formato en Clean Code ===");

const service = new ProductService();
const product = service.findProductByIdAndUpdateStockAndCalculateDiscountAndSendNotification(1, 5, 0.8);
console.log(product);

const processor = new OrderProcessor();
const items: OrderItem[] = [{ price: 100, quantity: 2 }];
processor.processOrder(1, items, "credit_card");

export { ProductService, OrderProcessor, UserManager, Product, Order, OrderItem, User };

File: katas/python/CleanCode/format/format_good.py
# Cumplimiento de Clean Code: Buen Formato
# ✅ Solución: Formato horizontal y vertical correcto

from typing import Dict, List, Any, Optional, TypedDict
from datetime import datetime


# ✅ Líneas cortas (idealmente max 80-100 caracteres)
# ✅ Espaciado apropiado entre operadores y elementos


class Product(TypedDict):
    id: int
    name: str
    price: float
    category: str
    stock: int
    supplier: str
    warranty: str
    discounted_price: Optional[float]


class ProductService:
    def __init__(self):
        self.products: List[Product] = [
            {
                "id": 1,
                "name": "Laptop",
                "price": 1200,
                "category": "Electronics",
                "stock": 50,
                "supplier": "TechCorp",
                "warranty": "2 years",
                "discounted_price": None,
            },
            {
                "id": 2,
                "name": "Mouse",
                "price": 25,
                "category": "Accessories",
                "stock": 200,
                "supplier": "TechCorp",
                "warranty": "1 year",
                "discounted_price": None,
            },
        ]

    # ✅ Líneas cortas, una acción por línea
    def find_product_by_id_and_update_stock_and_calculate_discount_and_send_notification(
        self, product_id: int, quantity: int, discount_rate: float
    ) -> Optional[Product]:
        product = self._find_product_by_id(product_id)

        if not product:
            return None

        product["stock"] -= quantity
        discounted_price = self._calculate_discounted_price(
            product["price"], discount_rate
        )

        print(
            f"Producto {product['name']} actualizado. "
            f"Stock: {product['stock']}, "
            f"Precio con descuento: {discounted_price}"
        )

        result = product.copy()
        result["discounted_price"] = discounted_price
        return result

    def _find_product_by_id(self, product_id: int) -> Optional[Product]:
        return next((p for p in self.products if p["id"] == product_id), None)

    def _calculate_discounted_price(self, price: float, discount_rate: float) -> float:
        return price * discount_rate


# ✅ Espaciado vertical apropiado
# ✅ Líneas en blanco separan conceptos relacionados
# ✅ Métodos relacionados están cerca


class OrderItem(TypedDict):
    price: float
    quantity: int


class Order(TypedDict):
    id: float
    customer_id: int
    items: List[OrderItem]
    total: float
    date: datetime
    status: str


class OrderProcessor:
    DISCOUNT_THRESHOLD = 100
    DISCOUNT_RATE = 0.9

    def __init__(self):
        self.orders: List[Order] = []

    # ✅ Grupo 1: Procesamiento de órdenes
    def process_order(
        self, customer_id: int, items: List[OrderItem], payment_method: str
    ) -> bool:
        if not self._is_valid_order(customer_id, items):
            return False

        total = self._calculate_total(items)
        order = self._create_order(customer_id, items, total)

        self._save_order(order)
        self._log_order_processed(total)

        return True

    def _is_valid_order(self, customer_id: int, items: List[OrderItem]) -> bool:
        return customer_id > 0 and items and len(items) > 0

    def _calculate_total(self, items: List[OrderItem]) -> float:
        total = sum(item["price"] * item["quantity"] for item in items)

        if total > self.DISCOUNT_THRESHOLD:
            total = total * self.DISCOUNT_RATE

        return total

    def _create_order(
        self, customer_id: int, items: List[OrderItem], total: float
    ) -> Order:
        return {
            "id": datetime.now().timestamp(),
            "customer_id": customer_id,
            "items": items,
            "total": total,
            "date": datetime.now(),
            "status": "pending",
        }

    def _save_order(self, order: Order) -> None:
        self.orders.append(order)

    def _log_order_processed(self, total: float) -> None:
        print(f"Orden procesada: ${total}")

    # ✅ Grupo 2: Consultas de órdenes (separado por línea en blanco)
    def get_orders(self) -> List[Order]:
        return self.orders

    # ✅ Grupo 3: Cancelación de órdenes (separado por línea en blanco)
    def cancel_order(self, order_id: float) -> bool:
        original_length = len(self.orders)
        self.orders = [o for o in self.orders if o["id"] != order_id]
        return len(self.orders) < original_length


# ✅ Formato consistente en toda la clase
# ✅ Espaciado uniforme, indentación correcta


class User(TypedDict):
    name: str
    email: str
    age: int
    created_at: datetime


class UserManager:
    def __init__(self):
        self.users: List[User] = []

    def add_user(self, name: str, email: str, age: int) -> User:
        user: User = {
            "name": name,
            "email": email,
            "age": age,
            "created_at": datetime.now(),
        }

        self.users.append(user)
        return user

    def find_user(self, email: str) -> Optional[User]:
        return next((u for u in self.users if u["email"] == email), None)

    def delete_user(self, email: str) -> None:
        self.users = [u for u in self.users if u["email"] != email]


if __name__ == "__main__":
    # Uso - También con buen formato
    print("=== Cumplimiento de Formato en Clean Code ===")

    service = ProductService()
    product = service.find_product_by_id_and_update_stock_and_calculate_discount_and_send_notification(
        1, 5, 0.8
    )
    print(product)

    processor = OrderProcessor()
    items: List[OrderItem] = [{"price": 100, "quantity": 2}]
    processor.process_order(1, items, "credit_card")
